src/middlewares/upload_logo.ts

```diff
@@ -0,0 +1,111 @@
+/**
+ * Middleware that get tasks performed to upload a logo
+ */
+import ApiManagementClient from "azure-arm-apimanagement";
+import {
+  IResponseSuccessRedirectToResource,
+  ResponseErrorForbiddenNotAuthorized,
+  ResponseErrorInternal,
+  ResponseErrorNotFound,
+  ResponseSuccessRedirectToResource
+} from "italia-ts-commons/lib/responses";
+import { NonEmptyString } from "italia-ts-commons/lib/strings";
+import {
+  getApimUser,
+  IExtendedUserContract,
+  isAdminUser
+} from "../apim_operations";
+import { AdUser } from "../bearer_strategy";
+import * as config from "../config";
+
+import { Logo as ApiLogo } from "../../generated/api/Logo";
+
+import {
+  fromLeft,
+  fromPredicate,
+  taskEither,
+  TaskEither,
+  tryCatch
+} from "fp-ts/lib/TaskEither";
+
+import { ErrorResponses, notificationApiClient } from "../controllers/services";
+
+import { toError } from "fp-ts/lib/Either";
+
+export const getApimUserTask = (
+  apiClient: ApiManagementClient,
+  authenticatedUser: AdUser
+): TaskEither<ErrorResponses, IExtendedUserContract> =>
+  tryCatch(
+    () => getApimUser(apiClient, authenticatedUser.emails[0]),
+    errors => ResponseErrorInternal(toError(errors).message)
+  ).foldTaskEither<ErrorResponses, IExtendedUserContract>(
+    error => fromLeft(error),
+    response =>
+      response.isSome()
+        ? taskEither.of(response.value)
+        : fromLeft(
+            ResponseErrorNotFound(
+              "API user not found",
+              "Cannot find a user in the API management with the provided email address"
+            )
+          )
+  );
+
+export const checkAdminTask = (
+  userAdmin: IExtendedUserContract
+): TaskEither<ErrorResponses, IExtendedUserContract> =>
+  fromPredicate(
+    (user: IExtendedUserContract) => isAdminUser(user),
+    _ => ResponseErrorForbiddenNotAuthorized
+  )(userAdmin);
+
+export const uploadServiceLogoTask = (
+  serviceId: NonEmptyString,
+  serviceLogo: ApiLogo
+): TaskEither<ErrorResponses, IResponseSuccessRedirectToResource<{}, {}>> =>
+  tryCatch(
+    () =>
+      notificationApiClient.uploadServiceLogo({
+        logo: serviceLogo,
+        serviceId
+      }),
+    errors => ResponseErrorInternal(toError(errors).message)
+  ).foldTaskEither(
+    error => fromLeft(error),
+    errorOrResponse =>
+      errorOrResponse && errorOrResponse.status === 201
```

Comment:
Could `errorOrResponse` be `undefined` ?

Reply:
Yes, can be undefined according to `uploadServiceLogo` defined in `api_client.ts.`

Reply:
Ok so if it can ben undefined, maybe we can wrap it into an option with `fromNullable` :)

Reply:
That's something that does not sound good to me. I think that the response body could be `undefined`, so the entire either `errorOrResponse` could not imho.

Reply:
The methods defined in `api-client.ts` return : 
- `Promise<IResponseType<404, Error, never> | ... 4 more ... | undefined>`
so the entire `errorOrResponse` can be undefined.
The code to avoid undefined could impact all methods of `api-client.ts` (maybe we could do it in a dedicated PR)

Reply:
Using io-ts the generated client map the same service as follows:
```
export type UploadServiceLogoT = r.IPutApiRequestType<
  {
    readonly service_id: string;
    readonly SubscriptionKey: string;
    readonly body: Logo;
  },
  "Content-Type" | "Ocp-Apim-Subscription-Key",
  never,
  | r.IResponseType<201, undefined, "Location">
  | r.IResponseType<400, ProblemJson, never>
  | r.IResponseType<401, undefined, never>
  | r.IResponseType<403, undefined, never>
  | r.IResponseType<404, undefined, never>
  | r.IResponseType<500, undefined, never>
>;
```
Can we replicate this?

Reply:
> Using io-ts the generated client map the same service as follows:
> 
> ```
> export type UploadServiceLogoT = r.IPutApiRequestType<
>   {
>     readonly service_id: string;
>     readonly SubscriptionKey: string;
>     readonly body: Logo;
>   },
>   "Content-Type" | "Ocp-Apim-Subscription-Key",
>   never,
>   | r.IResponseType<201, undefined, "Location">
>   | r.IResponseType<400, ProblemJson, never>
>   | r.IResponseType<401, undefined, never>
>   | r.IResponseType<403, undefined, never>
>   | r.IResponseType<404, undefined, never>
>   | r.IResponseType<500, undefined, never>
> >;
> ```
> 
> Can we replicate this?

ping :)

Reply:
To use this definition I have to  change `apiResponseDecodeer`, or maybe I create one to use only for new methods ?

Reply:
The first one :). Let's keep in mind that probably we should use` io-utils` to generate clients in this project cc @balanza @gunzip

Reply:
In this project there is already a client that is not generated with io-utils. I added the service to this but the problem is that in case of 400, the body is decoded as Error while in the model generated with io-utils is decoded as ProblemJson. A possible solution is to define ApiResponseType definition as follow:

```
export type ApiResponseType<R> =
  | IResponseType<200, R>
  | IResponseType<201, R, "Location">
  | IResponseType<400, ProblemJson | Error>
  | IResponseType<401, undefined | Error>
  | IResponseType<403, undefined | Error>
  | IResponseType<404, undefined | Error>
  | IResponseType<500, undefined | Error>;
```

Reply:
However `undefined` depends on how the client is defined in this project, there is a utility method that does a similar thing: https://github.com/pagopa/io-developer-portal-backend/blob/8b5ef26c95013f304bd9371ae84748dfdc2467a3/src/api_client.ts#L197

Reply:
Ok, for the moment let's use the utility to avoid directly check `undefined` response. Btw, when it's possible, we have to do another PR with io-utils upgrade 👍